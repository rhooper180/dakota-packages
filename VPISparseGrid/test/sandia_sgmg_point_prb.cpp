# include "sandia_rules.hpp"
# include "sandia_rules2.hpp"
# include "sandia_sgmg.hpp"

# include <cstdlib>
# include <iomanip>
# include <iostream>
# include <cmath>
//
//  Two global variables needed to support the "parameter" function.
//
double *P;
int *NP;

int main ( );

void sgmg_point_tests ( double tol );

void sgmg_point_test 
( 
  int dim_num,
  int level_max_min, 
  int level_max_max, 
  int growth, 
  void ( *gw_compute_points[] ) ( int order, int dim, double w[] ),
  double tol,
  int ( *gw_compute_order[] ) ( int level, int growth ) 
);

typedef void ( *GWPointer ) ( int order, int dim, double w[] );
typedef int ( *GWPointer2 ) ( int level, int growth );

//****************************************************************************80

int main ( )

//****************************************************************************80
//
//  Purpose:
//
//    MAIN tests SGMG_POINT.
//
//  Licensing:
//
//    This code is distributed under the GNU LGPL license.
//
//  Modified:
//
//    01 January 2011
//
//  Author:
//
//    John Burkardt
//
//  Reference:
//
//    Fabio Nobile, Raul Tempone, Clayton Webster,
//    A Sparse Grid Stochastic Collocation Method for Partial Differential
//    Equations with Random Input Data,
//    SIAM Journal on Numerical Analysis,
//    Volume 46, Number 5, 2008, pages 2309-2345.
//
{
  double tol;

  webbur::timestamp ( );
  std::cout << "\n";
  std::cout << "SGMG_POINT_PRB\n";
  std::cout << "  C++ version\n";
//
//  Generate the abscissas for sparse grid rules.
//
  tol = std::sqrt ( webbur::r8_epsilon ( ) );
  sgmg_point_tests ( tol );
//
//  Terminate.
//
  std::cout << "\n";
  std::cout << "SGMG_POINT_PRB\n";
  std::cout << "  Normal end of execution.\n";

  std::cout << "\n";
  webbur::timestamp ( );
  
  return 0;
}
namespace webbur 
{
//****************************************************************************80

double parameter ( int dim, int offset )

//****************************************************************************80
//
//  Purpose:
//
//    PARAMETER is a user-supplied routine to retrieve parameters.
//
//  Discussion:
//
//    The declaration for this function is in SANDIA_RULES.H
//
//  Licensing:
//
//    This code is distributed under the GNU LGPL license. 
//
//  Modified:
//
//    02 April 2010
//
//  Author:
//
//    John Burkardt
//
//  Parameters:
//
//    Input, int DIM, the spatial dimension.
//
//    Input, int OFFSET, the offset of the parameter within the 
//    spatial dimension.
//
//    Output, double PARAMETER, the value of the OFFSET-th parameter
//    associated with the DIM-th dimension.
//
{
  int i;
  int j;
  double value;

  j = 0;
  for ( i = 0; i < dim; i++ )
  {
    j = j + NP[i];
  }
  value = P[j+offset];

  return value;
}
}
//****************************************************************************80

void sgmg_point_tests ( double tol )

//****************************************************************************80
//
//  Purpose:
//
//    SGMG_POINT_TESTS calls SGMG_POINT_TEST.
//
//  Licensing:
//
//    This code is distributed under the GNU LGPL license.
//
//  Modified:
//
//    01 January 2011
//
//  Author:
//
//    John Burkardt
//
//  Parameters:
//
//    Input, double TOL, a tolerance for point equality.
//    A value of sqrt ( eps ) is reasonable, and will allow the code to
//    consolidate points which are equal, or very nearly so.  A value of
//    -1.0, on the other hand, will force the code to use every point, 
//    regardless of duplication.
//
{
  int dim_num;
  int growth;
  GWPointer2 *gw_compute_order;
  GWPointer *gw_compute_points;
  int level_max_max;
  int level_max_min;
  int np_sum;

  std::cout << "\n";
  std::cout << "SGMG_POINT_TESTS\n";
  std::cout << "  Call SGMG_POINT_TEST with various arguments.\n";
  std::cout << "\n";
  std::cout << "  All tests will use a point equality tolerance of " << tol << "\n";

  dim_num = 2;
  level_max_min = 0;
  level_max_max = 2;
  growth = 2;
  gw_compute_points = new GWPointer[dim_num];
  gw_compute_points[0] = webbur::clenshaw_curtis_points;
  gw_compute_points[1] = webbur::clenshaw_curtis_points;
  gw_compute_order = new GWPointer2[dim_num];
  gw_compute_order[0] = webbur::level_to_order_exp_cc;
  gw_compute_order[1] = webbur::level_to_order_exp_cc;
  NP = new int[dim_num];
  NP[0] = 0;
  NP[1] = 0;
  np_sum = webbur::i4vec_sum ( dim_num, NP );
  P = new double[np_sum];
  sgmg_point_test ( dim_num, level_max_min, level_max_max, growth, 
    gw_compute_points, tol, gw_compute_order );
  delete [] gw_compute_order;
  delete [] gw_compute_points;
  delete [] NP;
  delete [] P;

  dim_num = 2;
  level_max_min = 0;
  level_max_max = 2;
  growth = 2;
  gw_compute_points = new GWPointer[dim_num];
  gw_compute_points[0] = webbur::clenshaw_curtis_points;
  gw_compute_points[1] = webbur::patterson_points;
  gw_compute_order = new GWPointer2[dim_num];
  gw_compute_order[0] = webbur::level_to_order_exp_cc;
  gw_compute_order[1] = webbur::level_to_order_exp_gp;
  NP = new int[dim_num];
  NP[0] = 0;
  NP[1] = 0;
  np_sum = webbur::i4vec_sum ( dim_num, NP );
  P = new double[np_sum];
  sgmg_point_test ( dim_num, level_max_min, level_max_max, growth, 
    gw_compute_points, tol, gw_compute_order );
  delete [] gw_compute_order;
  delete [] gw_compute_points;
  delete [] NP;
  delete [] P;

  dim_num = 2;
  level_max_min = 0;
  level_max_max = 2;
  growth = 1;
  gw_compute_points = new GWPointer[dim_num];
  gw_compute_points[0] = webbur::clenshaw_curtis_points;
  gw_compute_points[1] = webbur::legendre_points;
  gw_compute_order = new GWPointer2[dim_num];
  gw_compute_order[0] = webbur::level_to_order_exp_cc;
  gw_compute_order[1] = webbur::level_to_order_linear_wn;
  NP = new int[dim_num];
  NP[0] = 0;
  NP[1] = 0;
  np_sum = webbur::i4vec_sum ( dim_num, NP );
  P = new double[np_sum];
  sgmg_point_test ( dim_num, level_max_min, level_max_max, growth, 
    gw_compute_points, tol, gw_compute_order );
  delete [] gw_compute_order;
  delete [] gw_compute_points;
  delete [] NP;
  delete [] P;

  dim_num = 2;
  level_max_min = 0;
  level_max_max = 2;
  growth = 1;
  gw_compute_points = new GWPointer[dim_num];
  gw_compute_points[0] = webbur::clenshaw_curtis_points;
  gw_compute_points[1] = webbur::laguerre_points;
  gw_compute_order = new GWPointer2[dim_num];
  gw_compute_order[0] = webbur::level_to_order_exp_cc;
  gw_compute_order[1] = webbur::level_to_order_linear_nn;
  NP = new int[dim_num];
  NP[0] = 0;
  NP[1] = 0;
  np_sum = webbur::i4vec_sum ( dim_num, NP );
  P = new double[np_sum];
  sgmg_point_test ( dim_num, level_max_min, level_max_max, growth, 
    gw_compute_points, tol, gw_compute_order );
  delete [] gw_compute_order;
  delete [] gw_compute_points;
  delete [] NP;
  delete [] P;

  dim_num = 2;
  level_max_min = 0;
  level_max_max = 2;
  growth = 1;
  gw_compute_points = new GWPointer[dim_num];
  gw_compute_points[0] = webbur::clenshaw_curtis_points;
  gw_compute_points[1] = webbur::gen_laguerre_points;
  gw_compute_order = new GWPointer2[dim_num];
  gw_compute_order[0] = webbur::level_to_order_exp_cc;
  gw_compute_order[1] = webbur::level_to_order_linear_nn;
  NP = new int[dim_num];
  NP[0] = 0;
  NP[1] = 1;
  np_sum = webbur::i4vec_sum ( dim_num, NP );
  P = new double[np_sum];
  P[0] = 1.5;
  sgmg_point_test ( dim_num, level_max_min, level_max_max, growth, 
    gw_compute_points, tol, gw_compute_order );
  delete [] gw_compute_order;
  delete [] gw_compute_points;
  delete [] NP;
  delete [] P;

  dim_num = 2;
  level_max_min = 0;
  level_max_max = 2;
  growth = 1;
  gw_compute_points = new GWPointer[dim_num];
  gw_compute_points[0] = webbur::fejer2_points;
  gw_compute_points[1] = webbur::jacobi_points;
  gw_compute_order = new GWPointer2[dim_num];
  gw_compute_order[0] = webbur::level_to_order_exp_f2;
  gw_compute_order[1] = webbur::level_to_order_linear_nn;
  NP = new int[dim_num];
  NP[0] = 0;
  NP[1] = 2;
  np_sum = webbur::i4vec_sum ( dim_num, NP );
  P = new double[np_sum];
  P[0] = 0.5;
  P[1] = 1.5;
  sgmg_point_test ( dim_num, level_max_min, level_max_max, growth, 
    gw_compute_points, tol, gw_compute_order );
  delete [] gw_compute_order;
  delete [] gw_compute_points;
  delete [] NP;
  delete [] P;

  dim_num = 2;
  level_max_min = 0;
  level_max_max = 2;
  growth = 1;
  gw_compute_points = new GWPointer[dim_num];
  gw_compute_points[0] = webbur::gen_hermite_points;
  gw_compute_points[1] = webbur::hermite_genz_keister_points;
  gw_compute_order = new GWPointer2[dim_num];
  gw_compute_order[0] = webbur::level_to_order_linear_wn;
  gw_compute_order[1] = webbur::level_to_order_exp_hgk;
  NP = new int[dim_num];
  NP[0] = 1;
  NP[1] = 0;
  np_sum = webbur::i4vec_sum ( dim_num, NP );
  P = new double[np_sum];
  P[0] = 2.0;
  sgmg_point_test ( dim_num, level_max_min, level_max_max, growth, 
    gw_compute_points, tol, gw_compute_order );
  delete [] gw_compute_order;
  delete [] gw_compute_points;
  delete [] NP;
  delete [] P;

  dim_num = 3;
  level_max_min = 0;
  level_max_max = 2;
  growth = 1;
  gw_compute_points = new GWPointer[dim_num];
  gw_compute_points[0] = webbur::clenshaw_curtis_points;
  gw_compute_points[1] = webbur::legendre_points;
  gw_compute_points[2] = webbur::hermite_points;
  gw_compute_order = new GWPointer2[dim_num];
  gw_compute_order[0] = webbur::level_to_order_exp_cc;
  gw_compute_order[1] = webbur::level_to_order_linear_wn;
  gw_compute_order[2] = webbur::level_to_order_linear_wn;
  NP = new int[dim_num];
  NP[0] = 0;
  NP[1] = 0;
  NP[2] = 0;
  np_sum = webbur::i4vec_sum ( dim_num, NP );
  P = new double[np_sum];
  sgmg_point_test ( dim_num, level_max_min, level_max_max, growth, 
    gw_compute_points, tol, gw_compute_order );
  delete [] gw_compute_order;
  delete [] gw_compute_points;
  delete [] NP;
  delete [] P;
//
//  Dimension 2, Level 3, Rule 3, Slow Exponential Growth.
//
  dim_num = 2;
  level_max_min = 3;
  level_max_max = 3;
  growth = 0;
  gw_compute_points = new GWPointer[dim_num];
  gw_compute_points[0] = webbur::patterson_points;
  gw_compute_points[1] = webbur::patterson_points;
  gw_compute_order = new GWPointer2[dim_num];
  gw_compute_order[0] = webbur::level_to_order_exp_gp;
  gw_compute_order[1] = webbur::level_to_order_exp_gp;
  NP = new int[dim_num];
  NP[0] = 0;
  NP[1] = 0;
  np_sum = webbur::i4vec_sum ( dim_num, NP );
  P = new double[np_sum];
  sgmg_point_test ( dim_num, level_max_min, level_max_max, growth, 
    gw_compute_points, tol, gw_compute_order );
  delete [] gw_compute_order;
  delete [] gw_compute_points;
  delete [] NP;
  delete [] P;
//
//  Dimension 2, Level 3, Rule 3, Moderate Exponential Growth.
//
  dim_num = 2;
  level_max_min = 3;
  level_max_max = 3;
  growth = 1;
  gw_compute_points = new GWPointer[dim_num];
  gw_compute_points[0] = webbur::patterson_points;
  gw_compute_points[1] = webbur::patterson_points;
  gw_compute_order = new GWPointer2[dim_num];
  gw_compute_order[0] = webbur::level_to_order_exp_gp;
  gw_compute_order[1] = webbur::level_to_order_exp_gp;
  NP = new int[dim_num];
  NP[0] = 0;
  NP[1] = 0;
  np_sum = webbur::i4vec_sum ( dim_num, NP );
  P = new double[np_sum];
  sgmg_point_test ( dim_num, level_max_min, level_max_max, growth, 
    gw_compute_points, tol, gw_compute_order );
  delete [] gw_compute_order;
  delete [] gw_compute_points;
  delete [] NP;
  delete [] P;

  return;
}
//***************************************************************************80

void sgmg_point_test 
(
  int dim_num,
  int level_max_min,
  int level_max_max, 
  int growth, 
  void ( *gw_compute_points[] ) ( int order, int dim, double w[] ),
  double tol,
  int ( *gw_compute_order[] ) ( int level, int growth )
)

//***************************************************************************80
//
//  Purpose:
//
//    SGMG_POINT_TEST tests SGMG_POINT.
//
//  Licensing:
//
//    This code is distributed under the GNU LGPL license.
//
//  Modified:
//
//    01 January 2011
//
//  Author:
//
//    John Burkardt
//
//  Parameters:
//
//    Input, int DIM_NUM, the spatial dimension.
//
//    Input, int LEVEL_MAX_MIN, LEVEL_MAX_MAX, the minimum and
//    maximum values of LEVEL_MAX.
//
//    Input, int GROWTH, the growth rule. 
//    0, slow;
//    1, moderate;
//    2, full.
//
//    Input, void ( *GW_COMPUTE_POINTS[] ) ( int order, int dim, double x[] ),
//    an array of pointers to functions which return the 1D quadrature points 
//    associated with each spatial dimension.
//
//    Input, double TOL, a tolerance for point equality.
//
//    Input, int ( *GW_COMPUTE_ORDER[] ) ( int level, int growth ),
//    an array of pointers to functions which return the order of the
//    1D quadrature rule of a given level and growth rule.
//
{
  int dim;
  int level_max;
  int point;
  int point_num;
  int point_total_num;
  int *sparse_index;
  int *sparse_order;
  double *sparse_point;
  int *sparse_unique_index;

  std::cout << "\n";
  std::cout << "SGMG_POINT_TEST\n";
  std::cout << "  SGMG_POINT returns an array of the points\n";
  std::cout << "  forming a multidimensional sparse grid with mixed factors.\n";
  std::cout << "\n";
  std::cout << "  Each sparse grid is of spatial dimension DIM_NUM,\n";
  std::cout << "  and is made up of product grids of levels up to LEVEL_MAX.\n";

  for ( level_max = level_max_min; level_max <= level_max_max; level_max++ )
  {
    point_total_num = webbur::sgmg_size_total ( dim_num, level_max, growth,
      gw_compute_order );

    point_num = webbur::sgmg_size ( dim_num, level_max, gw_compute_points, tol, 
      growth, gw_compute_order );

    sparse_unique_index = new int[point_total_num];

    webbur::sgmg_unique_index ( dim_num, level_max, gw_compute_points, tol, 
      point_num, point_total_num, growth, gw_compute_order, sparse_unique_index );

    sparse_order = new int[dim_num*point_num];
    sparse_index = new int[dim_num*point_num];

    webbur::sgmg_index ( dim_num, level_max, point_num, point_total_num, 
      sparse_unique_index, growth, gw_compute_order, sparse_order, sparse_index );

    sparse_point = new double [ dim_num * point_num ];

    webbur::sgmg_point ( dim_num, level_max, gw_compute_points, point_num, 
      sparse_order, sparse_index, growth, gw_compute_order, sparse_point );

    std::cout << "\n";
    std::cout << "  For LEVEL_MAX = " << level_max << "\n";
    std::cout << "\n";
    for ( point = 0; point < point_num; point++ )
    {
      std::cout << "  " << std::setw(4) << point << "  ";
      for ( dim = 0; dim < dim_num; dim++ )
      {
        std::cout << "  " << std::setw(10) << sparse_point[dim+point*dim_num];
      }
      std::cout << "\n";
    }
    delete [] sparse_index;
    delete [] sparse_order;
    delete [] sparse_point;
    delete [] sparse_unique_index;
  }
  return;
}

