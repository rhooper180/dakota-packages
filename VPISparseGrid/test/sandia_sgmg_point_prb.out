13 January 2012 10:41:19 AM

SGMG_POINT_PRB
  C++ version

SGMG_POINT_TESTS
  Call SGMG_POINT_TEST with various arguments.

  All tests will use a point equality tolerance of 1.49012e-08

SGMG_POINT_TEST
  SGMG_POINT returns an array of the points
  forming a multidimensional sparse grid with mixed factors.

  Each sparse grid is of spatial dimension DIM_NUM,
  and is made up of product grids of levels up to LEVEL_MAX.

  For LEVEL_MAX = 0

     0             0           0

  For LEVEL_MAX = 1

     0            -1           0
     1             0          -1
     2             0           0
     3             0           1
     4             1           0

  For LEVEL_MAX = 2

     0            -1          -1
     1            -1           0
     2            -1           1
     3     -0.707107           0
     4             0          -1
     5             0   -0.707107
     6             0           0
     7             0    0.707107
     8             0           1
     9      0.707107           0
    10             1          -1
    11             1           0
    12             1           1

SGMG_POINT_TEST
  SGMG_POINT returns an array of the points
  forming a multidimensional sparse grid with mixed factors.

  Each sparse grid is of spatial dimension DIM_NUM,
  and is made up of product grids of levels up to LEVEL_MAX.

  For LEVEL_MAX = 0

     0             0           0

  For LEVEL_MAX = 1

     0            -1           0
     1             0   -0.774597
     2             0           0
     3             0    0.774597
     4             1           0

  For LEVEL_MAX = 2

     0            -1   -0.774597
     1            -1           0
     2            -1    0.774597
     3     -0.707107           0
     4             0   -0.960491
     5             0   -0.774597
     6             0   -0.434244
     7             0           0
     8             0    0.434244
     9             0    0.774597
    10             0    0.960491
    11      0.707107           0
    12             1   -0.774597
    13             1           0
    14             1    0.774597

SGMG_POINT_TEST
  SGMG_POINT returns an array of the points
  forming a multidimensional sparse grid with mixed factors.

  Each sparse grid is of spatial dimension DIM_NUM,
  and is made up of product grids of levels up to LEVEL_MAX.

  For LEVEL_MAX = 0

     0             0           0

  For LEVEL_MAX = 1

     0            -1           0
     1     -0.707107           0
     2             0   -0.774597
     3             0  1.99493e-17
     4             0    0.774597
     5      0.707107           0
     6             1           0

  For LEVEL_MAX = 2

     0            -1   -0.774597
     1            -1  1.99493e-17
     2            -1    0.774597
     3      -0.92388           0
     4     -0.707107   -0.774597
     5     -0.707107  1.99493e-17
     6     -0.707107    0.774597
     7     -0.382683           0
     8             0    -0.90618
     9             0   -0.774597
    10             0   -0.538469
    11             0  -1.08185e-16
    12             0    0.538469
    13             0    0.774597
    14             0     0.90618
    15      0.382683           0
    16      0.707107   -0.774597
    17      0.707107  1.99493e-17
    18      0.707107    0.774597
    19       0.92388           0
    20             1   -0.774597
    21             1  1.99493e-17
    22             1    0.774597

SGMG_POINT_TEST
  SGMG_POINT returns an array of the points
  forming a multidimensional sparse grid with mixed factors.

  Each sparse grid is of spatial dimension DIM_NUM,
  and is made up of product grids of levels up to LEVEL_MAX.

  For LEVEL_MAX = 0

     0             0           1

  For LEVEL_MAX = 1

     0            -1           1
     1     -0.707107           1
     2             0    0.415775
     3             0           1
     4             0     2.29428
     5             0     6.28995
     6      0.707107           1
     7             1           1

  For LEVEL_MAX = 2

     0            -1    0.415775
     1            -1           1
     2            -1     2.29428
     3            -1     6.28995
     4      -0.92388           1
     5     -0.707107    0.415775
     6     -0.707107           1
     7     -0.707107     2.29428
     8     -0.707107     6.28995
     9     -0.382683           1
    10             0     0.26356
    11             0    0.415775
    12             0           1
    13             0      1.4134
    14             0     2.29428
    15             0     3.59643
    16             0     6.28995
    17             0     7.08581
    18             0     12.6408
    19      0.382683           1
    20      0.707107    0.415775
    21      0.707107           1
    22      0.707107     2.29428
    23      0.707107     6.28995
    24       0.92388           1
    25             1    0.415775
    26             1           1
    27             1     2.29428
    28             1     6.28995

SGMG_POINT_TEST
  SGMG_POINT returns an array of the points
  forming a multidimensional sparse grid with mixed factors.

  Each sparse grid is of spatial dimension DIM_NUM,
  and is made up of product grids of levels up to LEVEL_MAX.

  For LEVEL_MAX = 0

     0             0         2.5

  For LEVEL_MAX = 1

     0            -1         2.5
     1     -0.707107         2.5
     2             0      1.2204
     3             0         2.5
     4             0     3.80888
     5             0     8.47072
     6      0.707107         2.5
     7             1         2.5

  For LEVEL_MAX = 2

     0            -1      1.2204
     1            -1         2.5
     2            -1     3.80888
     3            -1     8.47072
     4      -0.92388         2.5
     5     -0.707107      1.2204
     6     -0.707107         2.5
     7     -0.707107     3.80888
     8     -0.707107     8.47072
     9     -0.382683         2.5
    10             0    0.817632
    11             0      1.2204
    12             0     2.47233
    13             0         2.5
    14             0     3.80888
    15             0     5.11601
    16             0     8.47072
    17             0     9.04415
    18             0     15.0499
    19      0.382683         2.5
    20      0.707107      1.2204
    21      0.707107         2.5
    22      0.707107     3.80888
    23      0.707107     8.47072
    24       0.92388         2.5
    25             1      1.2204
    26             1         2.5
    27             1     3.80888
    28             1     8.47072

SGMG_POINT_TEST
  SGMG_POINT returns an array of the points
  forming a multidimensional sparse grid with mixed factors.

  Each sparse grid is of spatial dimension DIM_NUM,
  and is made up of product grids of levels up to LEVEL_MAX.

  For LEVEL_MAX = 0

     0             0        0.25

  For LEVEL_MAX = 1

     0      -0.92388        0.25
     1     -0.707107        0.25
     2     -0.382683        0.25
     3             0   -0.537986
     4             0    0.152829
     5             0        0.25
     6             0    0.760157
     7      0.382683        0.25
     8      0.707107        0.25
     9       0.92388        0.25

  For LEVEL_MAX = 2

     0     -0.980785        0.25
     1      -0.92388   -0.537986
     2      -0.92388    0.152829
     3      -0.92388        0.25
     4      -0.92388    0.760157
     5      -0.83147        0.25
     6     -0.707107   -0.537986
     7     -0.707107    0.152829
     8     -0.707107        0.25
     9     -0.707107    0.760157
    10      -0.55557        0.25
    11     -0.382683   -0.537986
    12     -0.382683    0.152829
    13     -0.382683        0.25
    14     -0.382683    0.760157
    15      -0.19509        0.25
    16             0   -0.769541
    17             0   -0.537986
    18             0   -0.370814
    19             0    0.110027
    20             0    0.152829
    21             0        0.25
    22             0    0.562906
    23             0    0.760157
    24             0    0.884088
    25       0.19509        0.25
    26      0.382683   -0.537986
    27      0.382683    0.152829
    28      0.382683        0.25
    29      0.382683    0.760157
    30       0.55557        0.25
    31      0.707107   -0.537986
    32      0.707107    0.152829
    33      0.707107        0.25
    34      0.707107    0.760157
    35       0.83147        0.25
    36       0.92388   -0.537986
    37       0.92388    0.152829
    38       0.92388        0.25
    39       0.92388    0.760157
    40      0.980785        0.25

SGMG_POINT_TEST
  SGMG_POINT returns an array of the points
  forming a multidimensional sparse grid with mixed factors.

  Each sparse grid is of spatial dimension DIM_NUM,
  and is made up of product grids of levels up to LEVEL_MAX.

  For LEVEL_MAX = 0

     0             0           0

  For LEVEL_MAX = 1

     0      -1.58114           0
     1             0    -1.22474
     2             0           0
     3             0     1.22474
     4       1.58114           0

  For LEVEL_MAX = 2

     0       -2.3175           0
     1      -1.58114    -1.22474
     2      -1.58114           0
     3      -1.58114     1.22474
     4      -1.27639           0
     5             0    -2.95921
     6             0    -2.02323
     7             0   -0.524034
     8             0    0.524034
     9             0     1.22474
    10             0     2.02323
    11             0     2.95921
    12             0    -1.22474
    13             0           0
    14       1.27639           0
    15       1.58114    -1.22474
    16       1.58114           0
    17       1.58114     1.22474
    18        2.3175           0

SGMG_POINT_TEST
  SGMG_POINT returns an array of the points
  forming a multidimensional sparse grid with mixed factors.

  Each sparse grid is of spatial dimension DIM_NUM,
  and is made up of product grids of levels up to LEVEL_MAX.

  For LEVEL_MAX = 0

     0             0           0           0

  For LEVEL_MAX = 1

     0            -1           0           0
     1     -0.707107           0           0
     2             0   -0.774597           0
     3             0           0    -1.22474
     4             0           0  -9.86284e-17
     5             0           0     1.22474
     6             0    0.774597           0
     7      0.707107           0           0
     8             1           0           0

  For LEVEL_MAX = 2

     0            -1   -0.774597           0
     1            -1           0    -1.22474
     2            -1           0     1.22474
     3            -1           0  -9.86284e-17
     4            -1    0.774597           0
     5      -0.92388           0           0
     6     -0.707107   -0.774597           0
     7     -0.707107           0    -1.22474
     8     -0.707107           0  -9.86284e-17
     9     -0.707107           0     1.22474
    10     -0.707107    0.774597           0
    11     -0.382683           0           0
    12             0    -0.90618           0
    13             0   -0.774597    -1.22474
    14             0   -0.774597  -9.86284e-17
    15             0   -0.774597     1.22474
    16             0   -0.538469           0
    17             0           0    -2.02018
    18             0           0   -0.958572
    19             0           0    0.958572
    20             0  1.99493e-17     1.22474
    21             0           0     2.02018
    22             0  1.99493e-17    -1.22474
    23             0           0  2.40258e-16
    24             0    0.538469           0
    25             0    0.774597    -1.22474
    26             0    0.774597  -9.86284e-17
    27             0    0.774597     1.22474
    28             0     0.90618           0
    29      0.382683           0           0
    30      0.707107   -0.774597           0
    31      0.707107           0    -1.22474
    32      0.707107           0  -9.86284e-17
    33      0.707107           0     1.22474
    34      0.707107    0.774597           0
    35       0.92388           0           0
    36             1   -0.774597           0
    37             1           0    -1.22474
    38             1           0     1.22474
    39             1           0  -9.86284e-17
    40             1    0.774597           0

SGMG_POINT_TEST
  SGMG_POINT returns an array of the points
  forming a multidimensional sparse grid with mixed factors.

  Each sparse grid is of spatial dimension DIM_NUM,
  and is made up of product grids of levels up to LEVEL_MAX.

  For LEVEL_MAX = 3

     0     -0.960491           0
     1     -0.774597   -0.774597
     2     -0.774597           0
     3     -0.774597    0.774597
     4     -0.434244           0
     5             0   -0.960491
     6             0   -0.774597
     7             0   -0.434244
     8             0           0
     9             0    0.434244
    10             0    0.774597
    11             0    0.960491
    12      0.434244           0
    13      0.774597   -0.774597
    14      0.774597           0
    15      0.774597    0.774597
    16      0.960491           0

SGMG_POINT_TEST
  SGMG_POINT returns an array of the points
  forming a multidimensional sparse grid with mixed factors.

  Each sparse grid is of spatial dimension DIM_NUM,
  and is made up of product grids of levels up to LEVEL_MAX.

  For LEVEL_MAX = 3

     0     -0.993832           0
     1     -0.960491   -0.774597
     2     -0.960491           0
     3     -0.960491    0.774597
     4     -0.888459           0
     5     -0.774597   -0.960491
     6     -0.774597   -0.774597
     7     -0.774597   -0.434244
     8     -0.774597           0
     9     -0.774597    0.434244
    10     -0.774597    0.774597
    11     -0.774597    0.960491
    12     -0.621103           0
    13     -0.434244   -0.774597
    14     -0.434244           0
    15     -0.434244    0.774597
    16     -0.223387           0
    17             0   -0.993832
    18             0   -0.960491
    19             0   -0.888459
    20             0   -0.774597
    21             0   -0.621103
    22             0   -0.434244
    23             0   -0.223387
    24             0           0
    25             0    0.223387
    26             0    0.434244
    27             0    0.621103
    28             0    0.774597
    29             0    0.888459
    30             0    0.960491
    31             0    0.993832
    32      0.223387           0
    33      0.434244   -0.774597
    34      0.434244           0
    35      0.434244    0.774597
    36      0.621103           0
    37      0.774597   -0.960491
    38      0.774597   -0.774597
    39      0.774597   -0.434244
    40      0.774597           0
    41      0.774597    0.434244
    42      0.774597    0.774597
    43      0.774597    0.960491
    44      0.888459           0
    45      0.960491   -0.774597
    46      0.960491           0
    47      0.960491    0.774597
    48      0.993832           0

SGMG_POINT_PRB
  Normal end of execution.

13 January 2012 10:41:19 AM
